promql/engine_test.go

```diff
@@ -542,234 +711,496 @@ load 10s
+			End:        time.Unix(10, 0),
+			Interval:   5 * time.Second,
+		},
+		{
+			// Result is duplicated, so @ also produces 3 samples.
+			Query:      "metric @ 10",
 			MaxSamples: 3,
-			Result: Result{
-				nil,
-				Matrix{Series{
-					Points: []Point{{V: 1, T: 0}, {V: 1, T: 5000}, {V: 2, T: 10000}},
-					Metric: labels.FromStrings("__name__", "metric")},
-				},
-				nil,
-			},
-			Start:    time.Unix(0, 0),
-			End:      time.Unix(10, 0),
-			Interval: 5 * time.Second,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(10, 0),
+			Interval:   5 * time.Second,
 		},
 		{
-			Query:      "metric",
-			MaxSamples: 2,
-			Result: Result{
-				ErrTooManySamples(env),
-				nil,
-				nil,
-			},
-			Start:    time.Unix(0, 0),
-			End:      time.Unix(10, 0),
-			Interval: 5 * time.Second,
+			// With @ modifier, the result is duplicated for the time range, hence it counts.
+			// The selector itself counts does not count for any sample.
+			// During execution, the buffer per series maxes out at 2 samples and storing result
+			// per series for 1 evaluation takes 2 more samples. Hence 4 here. After discarding buffer and
+			// duplication of result, it is 6 at max.
+			Query:      `rate(bigmetric[10s] @ 10)`,
+			MaxSamples: 6,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(10, 0),
+			Interval:   5 * time.Second,
 		},
 		{
-			Query:      "rate(bigmetric[1s])",
-			MaxSamples: 1,
-			Result: Result{
-				nil,
-				Matrix{},
-				nil,
-			},
-			Start:    time.Unix(0, 0),
-			End:      time.Unix(10, 0),
-			Interval: 5 * time.Second,
+			// The peak samples in memory is during first evaluation:
+			//   - Subquery takes 22 samples, 11 for each bigmetric,
+			//   - Result is calculated per series where the series samples is buffered, hence 11 more here.
+			//   - The result of two series is added before the last series buffer is discarded, so 2 more here.
+			//   Hence at peak it is 22 (subquery) + 11 (buffer of a series) + 2 (result from 2 series).
+			// The subquery samples and the buffer is discarded before duplicating.
+			Query:      `rate(bigmetric[10s:1s] @ 10)`,
+			MaxSamples: 35,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(10, 0),
+			Interval:   5 * time.Second,
 		},
+		{
+			// Here the reasoning is same as above. But LHS and RHS are done one after another.
+			// So while one of them takes 35 samples at peak, we need to hold the 2 sample
+			// result of the other till then.
+			Query:      `rate(bigmetric[10s:1s] @ 10) + rate(bigmetric[10s:1s] @ 30)`,
+			MaxSamples: 37,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(10, 0),
+			Interval:   5 * time.Second,
+		},
+		{
+			// Sample as above but with only 1 part as step invariant.
+			// Here the peak is causes by the non-step invariant part as it touches more time range.
+			// Hence at peak it is 2*21 (subquery) + 11 (buffer of a series per evaluation) + 6 (result from 2 series at 3 eval times).
```

Comment:
This doesn't seem right, the two subqueries results are never in memory at the same time.

Also why 21 here when it's 22 above?

I'd expect this to be 6 + 22 + 11 + 2 = 41.

Reply:
This is actually calculating only for the subquery without @.

Since it's `rate(bigmetric[10s:1s])` for the query from 10s to 20s at 5s interval, the subquery has to effectively load the samples of 2 series at 1s interval from 0s to 20s (both inclusive), hence it is 2*21 for subquery.

During evaluation, since it per eval time per series, it has to load 11 samples in the buffer.

At the end, the 6 result samples before the subquery samples and the 11 samples buffer are discarded.

Hence 2*21 + 11 + 6.

There is an optimisation possible where we drop the samples from subquery once the series is evaluated instead of all at the, but thats a separate PR worthy. Still the peak would be >42 since evaluating subquery will take 42 samples upfront. (another optimisation is not the calculate it upfront and do lazy evaluation, but not sure if it's efficient)